### PR TITLE
Support qwen for pytorch engine

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support Qwen for pytorch engine

opencompass results:

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by jjjjohnson** [comment]
Looks like not support logn_attn yet
image

**Comment by RunningLeon** [comment]
@jjjjohnson hi, thanks for the reminder. Will update later.

File: lmdeploy/pytorch/supported_models.py [hunk 1]
```diff
@@ -88,5 +88,9 @@ def is_supported(model_path: str):
                 # baichuan-13B not supported by pytorch
                 if cfg.num_attention_heads == 40 and cfg.vocab_size == 64000:
                     support_by_torch = False
+            elif arch == 'QWenLMHeadModel':
```
Comment (grimoire) [inline lmdeploy/pytorch/supported_models.py:91]: Can vl be used as a chat model.

File: lmdeploy/pytorch/models/qwen.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                                          to_local=True)
+
+    @classmethod
+    def _distribute_output_fn(cls, outputs, device_mesh: DeviceMesh):
+        """Distribution output hook."""
+        dist.all_reduce(outputs)
+        return outputs
+
+
+class PatchedQWenBlock(nn.Module):
```
Comment (grimoire) [inline lmdeploy/pytorch/models/qwen.py:202]: Do not rewrite this module.

**Review by grimoire — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]

**Comment by jjjjohnson** [comment]
@RunningLeon  Look like there is en error in Qwen pytorch backend...
```
from lmdeploy.messages import PytorchEngineConfig, EngineGenerationConfig
from lmdeploy.pytorch.engine import Engine
from lmdeploy.serve.async_engine import AsyncEngine
import logging
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)



model_name='qwen-14b'
engine_config = PytorchEngineConfig(model_name=model_name,
                                    tp=1,max_batch_size=8,
                                    adapters=None,num_gpu_blocks=200, num_cpu_blocks=100)


engin= AsyncEngine(model_path=model_path, backend='pytorch', model_name=model_name, backend_config=engine_config)

prompts = 'Hello'
engin(prompts)
```
image

**Comment by RunningLeon** [comment]
@jjjjohnson Hi, thanks for your feedback. I'll try to reproduce the issue. Coming back to you soon.

**Comment by RunningLeon** [comment]
@jjjjohnson hi, lmdeploy had supported qwen with old  version. It seems you are running new version of Qwen and some rewriting failes in this case. Could try PR https://github.com/InternLM/lmdeploy/pull/1351? thanks for your understanding.

**Comment by jjjjohnson** [comment]
Cool! THX!

**Pull request merged**